crates/astria-sequencer/src/storage/stored_value.rs

```diff
@@ -0,0 +1,55 @@
+use std::fmt::{
+    self,
+    Display,
+    Formatter,
+};
+
+use astria_eyre::{
+    eyre::WrapErr as _,
+    Result,
+};
+use borsh::{
+    BorshDeserialize,
+    BorshSerialize,
+};
+
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub(crate) enum StoredValue<'a> {
+    Unit,
+    Address(crate::address::storage::Value<'a>),
+    Assets(crate::assets::storage::Value<'a>),
+    Accounts(crate::accounts::storage::Value),
+    Authority(crate::authority::storage::Value<'a>),
+    Sequence(crate::sequence::storage::Value),
+    Bridge(crate::bridge::storage::Value<'a>),
+    Ibc(crate::ibc::storage::Value<'a>),
+    App(crate::app::storage::Value<'a>),
+    Grpc(crate::grpc::storage::Value<'a>),
+}
+
+impl<'a> StoredValue<'a> {
+    pub(crate) fn serialize(&self) -> Result<Vec<u8>> {
+        borsh::to_vec(&self).wrap_err("failed to serialize stored value")
+    }
+
+    pub(crate) fn deserialize(bytes: &[u8]) -> Result<Self> {
+        borsh::from_slice(bytes).wrap_err("failed to deserialize stored value")
+    }
+}
+
+impl<'a> Display for StoredValue<'a> {
```

Comment:
What's this display impl used for? For some form of tracing?

Reply:
Yes, currently it's only used by the `TryFrom<StoredValue>` impls in the components' storage modules to provide info in the `bail!` message as to what unexpected storage enum variant was retrieved rather than the expected one.

I tried to keep the output info pretty concise.  A couple of example error logs look like
```
accounts stored value type mismatch: expected fee, found address stored value address prefix abc
accounts stored value type mismatch: expected fee, found assets stored value trace-prefixed denom a/long/path/to/denom
```

We probably don't need a lot of info there.  I could always remove this `Display` impl and all those in the components' storage modules too if you prefer?  Or I'm up for changing this how you see fit - even the `Debug` impl isn't generally too bad:

```
accounts stored value type mismatch: expected fee, found Address(AddressPrefix(AddressPrefix("abc")))
accounts stored value type mismatch: expected fee, found Assets(TracePrefixedDenom(TracePrefixedDenom { trace: [("a", "long"), ("path", "to")], base_denom: "denom" }))
```
If we go that route, we'd probably want to hand-roll `Debug` for some of the bigger structs like block header to avoid this becoming too large.  OTOH - we shouldn't ever see one of the type-mismatch errors in production.

Reply:
OK, with that context I think `Debug` is totally ok here. As you note, something really terrible would have happened if we see this, so turning on `Debug` is probably ok (and including it in some error).

Because of that I would even avoid handrolling and just go with the full derived `Debug` - after all, who knows in how far it would be helpful. Just toss the entire kitchensink at the user. :)

Reply:
Done in [44170ce](https://github.com/astriaorg/astria/pull/1492/commits/44170ceefc479c5f7839a388b322ac2412962df2).